lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/codeaction/MavenCodeActionParticipant.java

```diff
@@ -0,0 +1,55 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Red Hat Inc. and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.lemminx.extensions.maven.participants.codeaction;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.lemminx.extensions.contentmodel.participants.XMLSyntaxErrorCode;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionParticipant;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionRequest;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+
+public class MavenCodeActionParticipant implements ICodeActionParticipant {
```

Comment:
What's the benefit of this class? Why not registering directly MavenNoGrammarConstructor to the registry?

Reply:
There will be like 9 Code Actions added including `MavenNoGrammarConstraintsCodeAction`, so this is like a compound code action participant that is easy to use to setup and dispatch CA requests.

Reply:
what's the benefit over just adding 9 code actions participants directly? Of over adding only 1 participant that can produce different code actions?

Reply:
It''s actually the same. A Code Action (like MavenNoGrammarConstraintsCodeAction) implements the same ICodeActionParticipant interface as MavenCodeActionParticipant. When we receive a Code Action request we need to dispatch it to a proper Code Action - this is actually a work for MavenCodeActionParticipant.

A Code Action source may be pretty large, so keeping them all in one MavenCodeActionParticipant's class  doesn't seem rational. As far as I looked into the m2e-core IMarker quickfixes - this first No Grammar Constrain is the smallest one.

Also, any new Code Actions may be easily registered in MavenCodeActionParticipant without touching/breaking the existing ones.

Reply:
> When we receive a Code Action request we need to dispatch it to a proper Code Action - this is actually a work for MavenCodeActionParticipant.

Why not registering multiple participants to the MavenLemminxExtension directly? I don't get what this indirection provides.

> A Code Action source may be pretty large, so keeping them all in one MavenCodeActionParticipant's class doesn't seem rational. 

OK, so multiple participants is better.

> Also, any new Code Actions may be easily registered in MavenCodeActionParticipant without touching/breaking the existing ones.

Registering multiple participants is already possible without MavenCodeActionParticipant. Also, LemMinX may actually disable faulty individual participants in a future; grouping them is not ideal if this gets implemented, as 1 faulty extension will also disable others.

Reply:
While working on #365 and #366 I start thinking that MavenCodeActionProvider is a subject to be changed/removed. Problem is that for many kinds of diagnostics we may have to return a bunch of code actions, not the only one according to the error code <-> CA map. So, maybe it'll really be better to register code actions directly in LME registry.

Reply:
Still, what current problem MavenCodeActionParticipant resolves that wouldn't be resolved by directly registering participants?

Reply:
Agreed.